### DISCUSSION TITLE
About using transforms.inverse

### DISCUSSION THREAD

**Discussion by leeh43** [Project-MONAI/MONAI#2588] [Q&A]
Hi, thank you for the great work in monai. I am having a problem on using transforms.inverse in monai. As I want to resample back to the original space of the images (with the same size of original images) and output the image as nifti with its original resolution, a runtime error occurs when I used the transforms.inverse. What do you recommend for the next steps?

Here is the screenshot of the transforms and the inverse code:
![Screenshot from 2021-07-11 11-20-58](https://user-images.githubusercontent.com/54121206/125202652-1bf5b480-e23a-11eb-9bec-6d5e1edc970a.png)
![Screenshot from 2021-07-11 11-20-50](https://user-images.githubusercontent.com/54121206/125202660-1dbf7800-e23a-11eb-9a3d-59a0684ba3e5.png)

Thanks a lot!

**Comment by Nic-Ma** [comment]
Hi @leeh43 ,

Thanks for your interests and experiments with MONAI.
Could you please help share your error message?

Thanks.

**Comment by leeh43** [comment]
Hi Nic,

Here is the error message about the inverse transforms. Thanks a lot!
![image](https://user-images.githubusercontent.com/54121206/125214980-6f3e2600-e27f-11eb-8db9-836a47090bce.png)

**Comment by Nic-Ma** [comment]
I see your problem, the `inverse` function is not supposed to call in application directly.
You can use `Invertd` post processing transform or `BatchInverseTransform` API instead.
Refer to this tutorial if you are using MONAI v0.5.3:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py#L70

Thanks.

**Comment by leeh43** [comment]
I am currently using the latest version of the monai (0.6.0rc2). Meanwhile, I have another errror occurred after following the tutorial with the same test transformation above for input.
![image](https://user-images.githubusercontent.com/54121206/125231345-3793a600-e2a0-11eb-9e0e-d37b33da970a.png)

Here is my post transform for inverting it back to original space:
![image](https://user-images.githubusercontent.com/54121206/125231461-66118100-e2a0-11eb-9766-b7f2dd58992a.png)

What do you recommend for the next step?

Thanks a lot!

**Comment by Nic-Ma** [comment]
Hi @leeh43 ,

MONAI v0.6 adjusted the logic of postprocessing, we are still updating all the tutorials for MONAI v0.6, so I would suggest you to use 0.5.3 so far.
Could you please help try it again?

Thanks.

**Comment by leeh43** [comment]
Thank you for the recommendation. I have created a new environment and installed 0.5.3 for monai. It seems like there is an import error for Invertd function.
![image](https://user-images.githubusercontent.com/54121206/125239871-e8a13d00-e2ae-11eb-826a-b376e786f24c.png)

**Comment by Nic-Ma** [comment]
Hi @leeh43 ,

Sorry I forgot that `Invertd` was not included in v0.5.3 release, you can use the `BatchInverseTransform` API for your case. Just follow the `Batch inverse` section of this tutorial:
https://github.com/Project-MONAI/tutorials/blob/master/modules/inverse_transforms_and_test_time_augmentations.ipynb

Thanks.

**Comment by leeh43** [comment]
Hi Nic,

I am so sorry for so annoying about the errors. After I change it to v0.5.3 release, I cannot import BatchInverseTransform and also cannot import it with v0.6.0rc2. Here is the error attachment:
![Screenshot from 2021-07-12 10-30-46](https://user-images.githubusercontent.com/54121206/125315248-896a1980-e2fc-11eb-8421-1c80a04ca5ef.png)

**Comment by Nic-Ma** [comment]
Hi @leeh43 ,

Actually, I am sorry to lead you to the old version of MONAI because the examples and tutorials were not ready this morning.
But now we just released the tutorials for v0.6.0, so you can update to use MONAI v0.6.0 and refer to this example:
https://github.com/Project-MONAI/tutorials/blob/0.6.0/3d_segmentation/torch/unet_inference_dict.py

Thanks.

**Comment by leeh43** [comment]
Hi Nic,

Thanks a lot! it works now and I have one more small questions about generating multi-label segmentation map with the post_transform. For activation, I use softmax and normally we compute the max on the channel to output the multi-label segmentation, but how can I output a multi-label map with [1,2,3,4....,N] using post_transform?

Thanks a lot!

**Comment by Nic-Ma** [comment]
Hi @leeh43 ,

Sorry I don't quite understand your question, could you please help share an example program about your request?
Or @rijobro maybe you can help share some comments about the `multi-label map with [1,2,3,4....,N]`?

Thanks in advance.

**Comment by rijobro** [comment]
I think the question is saying that the output of the network will have the shape: `B, C=N, H, W, [D]`. I don't see why you can't omit the `max` (i.e., don't calculate the most-probable class), and keep the probabilities of all segmented classes.

**Comment by leeh43** [comment]
Hi Nic,

So currently, my goal is to output a segmentation map like this:
![monai_1](https://user-images.githubusercontent.com/54121206/125321687-925de980-e302-11eb-95bd-d65c61e5c84c.png)

However, the post_transform lead to the softmax mapping with multiple channels which is like this:
![monai_2](https://user-images.githubusercontent.com/54121206/125321760-a43f8c80-e302-11eb-8e5f-420c28c2f42a.png)

Can we compute the max with the post_transform?

**Comment by Nic-Ma** [comment]
Hi @leeh43 ,

I think you can try the `AsDiscreted` postprocessing transform for your case:
https://github.com/Project-MONAI/MONAI/blob/0.6.0/monai/transforms/post/dictionary.py#L124.

Thanks.
